pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java

```diff
@@ -438,14 +438,21 @@ private PublishRate publishRateInBroker(ServiceConfiguration config) {
         return new PublishRate(config.getMaxPublishRatePerTopicInMessages(), config.getMaxPublishRatePerTopicInBytes());
     }
 
-    protected boolean isProducersExceeded() {
+    protected boolean isProducersExceeded(Producer producer) {
+        if (isSystemTopic() || producer.isRemote()) {
+            return false;
+        }
         Integer maxProducers = topicPolicies.getMaxProducersPerTopic().get();
-        if (maxProducers > 0 && maxProducers <= producers.size()) {
+        if (maxProducers != null && maxProducers > 0 && maxProducers <= getUserCreatedProducerSize()) {
             return true;
         }
         return false;
     }
 
+    private long getUserCreatedProducerSize() {
+        return producers.values().stream().filter(p -> !(p.isRemote() || p.getTopic().isSystemTopic())).count();
+    }
```

Comment:
Why do we need this change? All the producers under a topic instance should only point to this topic, right?

Reply:
All the producers (including user-created producers, the system topic producers, and replicator producers) stored in `producers `.  So when checking the user-created producer size, we should filter system topic producers and replicator producers.

Reply:
Ok, but `p.getTopic().isSystemTopic()` is not required? Line 442 is already handled.

Reply:
fixed